lra/coordinator/server/src/main/java/io/helidon/lra/coordinator/Participant.java

```diff
@@ -0,0 +1,605 @@
+        for (AtomicInteger i = new AtomicInteger(0); i.getAndIncrement() < SYNCHRONOUS_RETRY_CNT;) {
+            if (!sendingStatus.compareAndSet(SendingStatus.NOT_SENDING, SendingStatus.SENDING)) return false;
+            LOGGER.log(Level.FINE, () -> "Sending complete, sync retry: " + i.get()
+                    + ", status: " + status.get().name()
+                    + " statusUri: " + getStatusURI().map(URI::toASCIIString).orElse(null));
+            WebClientResponse response = null;
+            try {
+                if (status.get().isFinal()) {
+                    return true;
+                    // call for client status only on retries and when status uri is known
+                } else if (!status.get().equals(Status.ACTIVE) && getStatusURI().isPresent()) {
+                    // If the participant does not support idempotency then it MUST be able to report its status
+                    // by annotating one of the methods with the @Status annotation which should report the status
+                    // in case we can't retrieve status from participant just retry n times
+                    ParticipantStatus reportedClientStatus = retrieveStatus(lra, Completing).orElse(null);
+                    if (reportedClientStatus == Completed) {
+                        LOGGER.log(Level.INFO, "Participant reports it is completed.");
+                        status.set(Status.COMPLETED);
+                        return true;
+                    } else if (reportedClientStatus == FailedToComplete) {
+                        LOGGER.log(Level.INFO, "Participant reports it failed to complete.");
+                        status.set(Status.FAILED_TO_COMPLETE);
+                        return true;
+                    } else if (reportedClientStatus == Active) {
+                        // last call didn't reach participant, try call again
+                    } else if (reportedClientStatus == Completing) {
+                        LOGGER.log(Level.INFO, "Participant reports it is still completing.");
+                        status.set(Status.CLIENT_COMPLETING);
+                        return false;
+                    } else if (remainingCloseAttempts.decrementAndGet() <= 0) {
+                        LOGGER.log(Level.INFO, "Participant didnt report final status after {0} status call retries.",
+                                new Object[] {RETRY_CNT});
+                        status.set(Status.FAILED_TO_COMPLETE);
+                        return true;
+                    } else {
+                        // Unknown status, lets try in next recovery cycle
+                        return false;
+                    }
+                }
+                response = getWebClient(endpointURI.get())
+                        .put()
+                        .headers(lra.headers())
+                        .submit(LRAStatus.Closed.name())
+                        .await(timeout, TimeUnit.MILLISECONDS);
```

Comment:
Can you explain why this is waiting? Couldn't this method be implemented as reactive?

Reply:
There is complicated timeout logic in retry mechanism, clean reactive approach would be using timeout operator, but that would just mean managing one more thread pool. Imho readability would suffer greatly with async recursion.